### ISSUE TITLE
[🐛BUG] full_sort_topk OutOfMemoryError

### ISSUE THREAD

**Issue by Escape142** [RUCAIBox/RecBole#1866]
Bug Report Summary:

I encountered two issues while using the MultiVAE model on various datasets, including Yelp, Amazon_cds, and Goodreads. The problems are as follows:

Out of Memory Error: I am facing a CUDA Out of Memory Error, which occurred in the get_relevant_ranks function.

Distributed Data Parallel Issue: Additionally, it seems that the Distributed Data Parallel feature is not being utilized despite being set in the configuration.

Bug Details:

Description:
While working with the MultiVAE model on multiple datasets, including Yelp, Amazon_cds, and Goodreads, I have encountered two significant issues that need to be resolved.

Out of Memory Error (OOME):

Location: The error occurred at line 267 in the get_relevant_ranks function.
Error Message: The error message reported a CUDA Out of Memory Error, indicating that it attempted to allocate 19.45 GiB of GPU memory.
Context: The GPU in question has a total capacity of 79.15 GiB, with 59.61 GiB already allocated, leaving 18.29 GiB of free memory. Additionally, PyTorch reserved 59.64 GiB in total. The error suggests that setting max_split_size_mb might help avoid fragmentation.
Distributed Data Parallel Not Utilized:

Issue: Despite having the Distributed Data Parallel option set in the configuration, it appears this feature is not being used.
Configuration Setting: The relevant configuration setting is as follows:

Environment Details:

**Desktop:**
Operating System: Linux (Please specify the exact distribution if relevant)
RecBole Version: 1.1.1
Python Version: 3.9.17
PyTorch Version: 2.0.1
cudatoolkit Version: [Please specify the version if applicable]

**Configuration:**
{'gpu_id': '1,2',
 'worker': 0,
 'use_gpu': True,
 'seed': 2020,
 'state': 'INFO',
 'reproducibility': True,
 'data_path': 'data/tmp/yelp',
 'checkpoint_dir': 'saved',
 'show_progress': True,
 'save_dataset': False,
 'dataset_save_path': None,
 'save_dataloaders': False,
 'dataloaders_save_path': None,
 'log_wandb': False,
 'wandb_project': 'recbole',
 'shuffle': True,
 'epochs': 300,
 'train_batch_size': 4096,
 'learner': 'adam',
 'learning_rate': 0.001,
 'train_neg_sample_args': {'distribution': 'none',
  'sample_num': 'none',
  'alpha': 'none',
  'dynamic': False,
  'candidate_num': 0},
 'eval_step': 1,
 'stopping_step': 10,
 'clip_grad_norm': None,
 'weight_decay': 0.0,
 'loss_decimal_place': 4,
 'require_pow': False,
 'enable_amp': False,
 'enable_scaler': False,
 'transform': None,
 'eval_args': {'split': {'RS': [0.8, 0.1, 0.1]},
  'group_by': 'user',
  'order': 'RO',
  'mode': 'full'},
 'repeatable': False,
 'metrics': ['Recall', 'MRR', 'NDCG', 'Hit', 'Precision'],
 'topk': [10],
 'valid_metric': 'MRR@10',
 'valid_metric_bigger': True,
 'eval_batch_size': 2097152,
 'metric_decimal_place': 4,
 'mlp_hidden_size': [600],
 'latent_dimension': 128,
 'dropout_prob': 0.5,
 'anneal_cap': 0.2,
 'total_anneal_steps': 200000,
 'field_separator': '\t',
 'seq_separator': ' ',
 'USER_ID_FIELD': 'user_id',
 'ITEM_ID_FIELD': 'item_id',
 'RATING_FIELD': 'rating',
 'TIME_FIELD': 'timestamp',
 'seq_len': None,
 'LABEL_FIELD': 'label',
 'threshold': None,
 'NEG_PREFIX': 'neg_',
 'numerical_features': [],
 'load_col': {'inter': ['user_id', 'item_id', 'rating']},
 'unload_col': None,
 'unused_col': None,
 'additional_feat_suffix': None,
 'rm_dup_inter': None,
 'val_interval': None,
 'filter_inter_by_user_or_item': True,
 'user_inter_num_interval': '[0,inf)',
 'item_inter_num_interval': '[0,inf)',
 'alias_of_user_id': None,
 'alias_of_item_id': None,
 'alias_of_entity_id': None,
 'alias_of_relation_id': None,
 'preload_weight': None,
 'normalize_field': None,
 'normalize_all': None,
 'discretization': None,
 'ITEM_LIST_LENGTH_FIELD': 'item_length',
 'LIST_SUFFIX': '_list',
 'MAX_ITEM_LIST_LENGTH': 50,
 'POSITION_FIELD': 'position_id',
 'HEAD_ENTITY_ID_FIELD': 'head_id',
 'TAIL_ENTITY_ID_FIELD': 'tail_id',
 'RELATION_ID_FIELD': 'relation_id',
 'ENTITY_ID_FIELD': 'entity_id',
 'kg_reverse_r': False,
 'entity_kg_num_interval': '[0,inf)',
 'relation_kg_num_interval': '[0,inf)',
 'benchmark_filename': ['train', 'val', 'test'],
 'MODEL_TYPE': <ModelType.GENERAL: 1>,
 'neg_sampling': None,
 'dataset': 'yelp',
 'model': 'MultiVAE',
 'MODEL_INPUT_TYPE': <InputType.PAIRWISE: 2>,
 'eval_type': <EvaluatorType.RANKING: 1>,
 'single_spec': True,
 'local_rank': 0,
 'device': device(type='cuda'),
 'eval_neg_sample_args': {'distribution': 'uniform', 'sample_num': 'none'}}

**Comment by chenyuwuxin** [comment]
@Escape142 Thanks for your attention to RecBole!
For the first issue, it is caused by insufficient GPU memory on your machine. You can either upgrade to a GPU with larger memory or reduce the size of your dataset. We recommend setting the k-core to filter out some irrelevant interactions from your dataset. You can refer to the configuration file below for guidance:
```
# dataset config
gpu_id: 1,2
field_separator: "\t"
seq_separator: " "
USER_ID_FIELD: user_id
ITEM_ID_FIELD: item_id
RATING_FIELD: rating
NEG_PREFIX: neg_
LABEL_FIELD: label
load_col:
    inter: [user_id, item_id, rating]

# data filtering for interactions
val_interval:
    rating: "[3,inf)"    
unused_col: 
    inter: [rating]

user_inter_num_interval: "[10,inf)"
item_inter_num_interval: "[10,inf)"

# training and evaluation
epochs: 500
train_batch_size: 4096
eval_batch_size: 40960000
valid_metric: NDCG@10
train_neg_sample_args: 
    distribution: uniform
    sample_num: 1
    dynamic: False

# model
embedding_size: 64
```
For the second issue, you must specify the --nproc arguments in the command, e.g., `python run_recbole.py --nproc=2`

**Comment by Escape142** [comment]
Thank you for getting back to me.

I employed an NVIDIA A100-SXM-80GB GPU to retrieve ranks using the "full_sort_topk()" function. However, the current solution only processes data in batches.

Regarding distributed data parallelism (ddp), I configured it with "--nproc=2," but I'm still facing an issue where calculations are being executed on a single GPU.

**Comment by chenyuwuxin** [comment]
@Escape142 
Due to the large number of items in the item set, tensor operations consume a significant amount of storage space. Therefore, we recommend using the k-core function to filter out some items. Alternatively, you can also try changing the evaluation method, such as using pop100 for negative sampling, e.g., setting:
```
eval_args:
    mode: uni100
```

For the second issue, is the 'gpu_id' correct? You may need to double-check if the gpu_id is correct (i.e., whether GPU 1 and 2 are functioning properly), or specify `--gpu_id=1,2` directly in the command line parameters.

**Comment by Escape142** [comment]
'gpu_id' is correct.
